### DISCUSSION TITLE
routes using getServerSideProps on firebase returns 500 internal server error deployed

### DISCUSSION THREAD

**Discussion by atco-jh** [vercel/next.js#47699] [Help] [answered]
### Verify canary release

- [x] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:29 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T8101
    Binaries:
      Node: 16.15.1
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: 6.11.0
    Relevant packages:
      next: 13.2.4
      eslint-config-next: 13.2.4
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Data fetching (gS(S)P, getInitialProps), Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue

https://github.com/atco-jh/nextjs-firebase-ssg-ssr-example

### To Reproduce

1. Create a nextJS app
2. Firebase init and Create a firebase function inside that serves the pages. 
3. Build a page that has `getServerSideProps` 
4. Deploy the app to firebase
5. Navigate to the SSR page 

Using the repo I made above: 
- Pull the simplified repo, build it and the functions inside, then deploy to firebase
- The `about` and `home` pages are marked as SSG, and `post/[id].tsx` is SSR
- Navigate to "post by XX` pages and see internal server error

### Describe the Bug

Any routes that use `getServerSideProps` returns a 500 Internal Server error on firebase hosting. 

My app is a mix of SSG and SSR. The SSG pages load fine, but the SSR paths return a 500 internal server error. 

The routes work fine locally, and using the firebase emulator, but not when deployed onto production. No logs appear on the function side. 

Alternatively, am I doing something wrong with my configuration? 

Related Issues: #17498  [#20641](https://github.com/vercel/next.js/discussions/20641)

### Expected Behavior

SSR routes to render and not return a 500 error. 

### How are you deploying your application? (if relevant)

firebase

**Comment by atco-jh** [comment, marked as answer]
For anyone wondering and getting this error, it turns out there's a bunch of issues and the example in the repo for firebase hosting is far from correct. Here is everything I had to correct. I'm going to fix my example repo, and then do a PR into the nextjs for a working example. 

Since my file structure was 
```
.
└── CODEBASE-NAME/
    ├── public
    ├── src/
    │   ├── pages/
    │   └── index.tsx
    ├── package.json
    ├── firebase.json
    ├── next.config.json
    └── .firebaserc

```
1. You need to enable the experimental frameworks in your CLI. So login `firebase login`, make sure you're in the root directory: `firebase experiments:enable webframeworks`
2. Ensure all .env variables are all uppercases and Underscores only, and start with `NEXT_PUBLIC` if you want to expose them
3. in `package.json` I had  a `build` script and it turns out I don't need to do this at all and removed it. This was also (somehow) critical in it deploying (would error out every other time). By having it in my package.json, I was using `build: next build` that was somehow overriding firebase's build command. But if I added anything extra `build: next build /src` it was ignored and built successfully, so I ended up removing this completely. 
4. **Don't create functions to serve the pages** - Seriously! Firebase creates the function that serves the pages for you, so you don't need to create your own function to serve the pages. Firebase will name it starting with `ssr` and then your firebase project name, all lowercase and together. Eg. `sample-project-dev` function will automatically be named `ssrsampleprojectdev`. This is done on deploy automatically, so you can either remove the `functions` section in the `firebase.json` and/or delete the user created functions.
5. in `firebase.json` , delete all mention of `functions` section since it's being created for you. Under hosting, remove `predeploy` instructions unless it's outside of a `npm install` and `npm build`.
6. In `firebase.json`, replace `"public": "public"` with `"source": "."` or firebase will not be able to detect that it's a next.js project
7. Not sure if needed, but in `firebase.json` I left in the `rewrites` field but updated it so it references the function, so overall, my hosting file looks like this: 

```
{
  "hosting": {
    "source": ".",
    "cleanUrls": true,
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**",
    ],
    "rewrites": [
      {
        "source": "**",
        "function": "ssrsampleprojectdev"
      },
      {
        "source": "**/**",
        "function": "ssrsampleprojectdev"
      }
    ],
    "headers": [
      {
        "source": "**",
        "headers": [
          {
            "key": "Strict-Transport-Security",
            "value": "max-age=63072000; includeSubDomains; preload"
          }
        ]
      }
    ]
  }
}
```

**Reply by JwillSDP** [reply]
I will try this out. I am having the same problem
